torch/csrc/jit/source_range.h

```diff
@@ -7,29 +7,108 @@
 namespace torch {
 namespace jit {
 
+// This structure contains information regarding the original source file
+// associated with a SourceRange. In particular, it records the original file
+// name, as well as the offset of source_ with respect to the original source
+// file.
+//
+// That is, if a SourceRange describes a location within a function, source_ 
+// will contain only the source code for that function, and not the whole file
+// due to the semantics of Python's inspect module. To recover original source
+// file line information, we store away file_ofset_, which is the line offset
+// of source_ with respect to the beginning of its source file.
+struct File {
+  File(
+      std::shared_ptr<std::string> source,
+      std::string filename,
+      size_t file_offset)
+      : source_(std::move(source)),
+        filename_(std::move(filename)),
+        file_offset_(file_offset) {
+    size_t pos = 0;
+    do {
+      line_len_prefix_sum_.push_back(pos);
+      pos++;
+    } while ((pos = source_->find('\n', pos)) != std::string::npos);
+  }
+
+  File(std::string source, std::string filename, size_t file_offset)
+      : File(
+            std::make_shared<std::string>(source),
+            std::move(filename),
+            file_offset) {}
+
+  // Given a line number (within source_), return the byte offset of the
+  // beginning of that line.
+  size_t offset_for_line(size_t line) const {
+    return line_len_prefix_sum_.at(line);
+  }
+
+  // Given a byte offset (within source_), return the line number that that
+  // byte resides on. Note that the returned line number is with respect to
+  // source_, not the original file. It is the caller's responsibility to add
+  // file_offset_
+  size_t lineno_for_offset(size_t offset) const {
+    size_t line = 0;
+    for (size_t i = 0; i < line_len_prefix_sum_.size(); ++i) {
+      if (line_len_prefix_sum_[i] > offset) {
+        line = i;
+        break;
+      }
+    }
+    return line;
+  }
+
+  // File line number for a given offset.
+  size_t file_lineno_for_offset(size_t offset) const {
+    return lineno_for_offset(offset) + file_offset_;
+  }
+
+  // Source text of the original callable
+  std::shared_ptr<std::string> source_;
```

Comment:
There is no need for this to be a shared_ptr anymore, right? Because we already have a `shared_ptr<File>`